### PR TITLE
[lit-next] Refactor reserved properties

### PR DESCRIPTION
**Pull request by kevinpschaaf**
Cleans our rollup/terser reserved property situation a bit:
* Renames `crossPackagePropertyMangles` to `stableProperties`, and adds some better descriptions and validation to the list
* Implements a convention that all `stableProperties` are prefixed with `_$`, so it's clear in the source code when a property requires version stability and needs to be treated carefully
* Adds `$private` helper API for all properties that were reserved only for use by lit-ssr; the only `reservedProperties` are now the ones used for branding `TemplateResult` and `Directive`
  * This adde a `$private` export and `private-ssr-support.ts` module to `lit-element` as well, to expose `attributeToProperty`; it's a bit marginal, but follows the same pattern as the other packages that need it
* Adds `resetPartValue` helper for use by directives that were accessing the private `_value` property

### CODE REVIEW

File: rollup-common.js [hunk 1]
```diff
@@ -40,54 +40,75 @@ const skipBundleOutput = {
...
+// they are easily identifiable as properties requiring version stability and
+// thus special attention.
+//
+// Mangled names are uppercase letters, in case we ever might want to use
+// lowercase letters for short, public APIs. Keep this list in order by mangled
+// name to avoid accidental re-assignments. When adding a name, add to the end
+// and choose the next letter.
+//
+// ONCE A MANGLED NAME HAS BEEN ASSIGNED TO A PROPERTY, IT MUST NEVER BE USED
+// FOR A DIFFERENT PROPERTY IN SUBSEQUENT VERSIONS.
```
Comment (kevinpschaaf) [inline rollup-common.js:64]: I'm obv breaking this rule here once to get the list in order; after this we should follow the rule and not have gaps in the list, since it was getting pretty confusing.

File: packages/lit-element/src/lit-element.ts [hunk 2]
```diff
@@ -170,3 +170,13 @@ if (DEV_MODE) {
     return true;
   };
 }
+
+export const $private = {
```
Comment (sorvell) [inline packages/lit-element/src/lit-element.ts:174]: Few questions:

* Why is this here and not in UpdatingElement since `_$attributeToProperty` is defined there? Is it because we only support SSR for LitElement?
* What does the leading `$` indicate here? 
* If this is for SSR, should it be named as such?

File: packages/lit-element/src/private-ssr-support.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {$private as p} from './lit-element.js';
```
Comment (sorvell) [inline packages/lit-element/src/private-ssr-support.ts:15]: I realize you tried to explain this in the PR comments, but I think this module needs some comments or we'll forget why it's being done this circuitous way.

File: packages/lit-html/src/directives/live.ts [hunk 4]
```diff
@@ -25,10 +25,7 @@ import {
   BOOLEAN_ATTRIBUTE_PART,
   ATTRIBUTE_PART,
 } from '../lit-html.js';
-
-// A sentinal value that can never appear as a part value except when set by
-// live(). Used to force a dirty-check to fail and cause a re-render.
-const RESET_VALUE = {};
+import {resetPartValue} from '../parts.js';
```
Comment (sorvell) [inline packages/lit-html/src/directives/live.ts:28]: What does this have to do with reserved properties?

File: packages/lit-html/src/parts.ts [hunk 5]
```diff
@@ -76,35 +76,43 @@ export const setPartValue = <T extends Part>(
...
 };
 
-export const getPartValue = (part: NodePart) => part._value;
+export const getPartValue = (part: NodePart) => part._$value;
+
+// A sentinal value that can never appear as a part value except when set by
+// live(). Used to force a dirty-check to fail and cause a re-render.
+const RESET_VALUE = {};
+
+export const resetPartValue = (part: Part, value: unknown = RESET_VALUE) =>
```
Comment (sorvell) [inline packages/lit-html/src/parts.ts:94]: Why is this here? Add to PR description?

File: packages/lit-html/src/private-ssr-support.ts [hunk 6]
```diff
@@ -29,4 +35,22 @@ export const $private = {
   markerMatch: p._markerMatch,
   HTML_RESULT: p._HTML_RESULT,
   getTemplateHtml: p._getTemplateHtml,
+  patchDirectiveResolve: (
```
Comment (sorvell) [inline packages/lit-html/src/private-ssr-support.ts:38]: Why were these added?

File: rollup-common.js [hunk 7]
```diff
@@ -40,54 +40,75 @@ const skipBundleOutput = {
...
+const stableProperties = {
```
Comment (sorvell) [inline rollup-common.js:65]: I think we should try to add better descriptions for each property as to why it's here.

**Review by sorvell — CHANGES_REQUESTED** [review]

File: packages/lit-html/src/parts.ts (hunk 5 above)
Comment (kevinpschaaf) [reply packages/lit-html/src/parts.ts:94]: Added to PR description:
> * Adds `resetPartValue` helper for use by directives that were accessing the private `_value` property

File: packages/lit-html/src/directives/live.ts (hunk 4 above)
Comment (kevinpschaaf) [reply packages/lit-html/src/directives/live.ts:28]: Added to PR description:
> * Adds `resetPartValue` helper for use by directives that were accessing the private `_value` property

File: packages/lit-element/src/private-ssr-support.ts (hunk 3 above)
Comment (kevinpschaaf) [reply packages/lit-element/src/private-ssr-support.ts:15]: Whoops, this is intended to just copy the pattern already established in `lit-html`; I added the comment from the same file in `lit-html`:
```js
/**
 * END USERS SHOULD NOT RELY ON THIS OBJECT.
 *
 * We currently do not make a mangled rollup build of the lit-ssr code. In order
 * to keep a number of (otherwise private) top-level exports  mangled in the
 * client side code, we export a $private object containing those members (or
 * helper methods for accessing private fields of those members), and then
 * re-export them for use in lit-ssr. This keeps lit-ssr agnostic to whether the
 * client-side code is being used in `dev` mode or `prod` mode.
 */
```

File: packages/lit-element/src/lit-element.ts (hunk 2 above)
Comment (sorvell) [reply packages/lit-element/src/lit-element.ts:187]: OK, I understand this based on the pattern done in lit-html

File: packages/lit-element/src/private-ssr-support.ts (hunk 3 above)
Comment (sorvell) [reply packages/lit-element/src/private-ssr-support.ts:15]: Let's try to change this to `_private` instead of `$private` to match our convention for private (assuming top level exports don't get mangled).

File: packages/lit-html/src/private-ssr-support.ts (hunk 6 above)
Comment (sorvell) [reply packages/lit-html/src/private-ssr-support.ts:39]: Its' to avoid use of mangled properties.

File: packages/lit-element/src/lit-element.ts (hunk 2 above)
Comment (kevinpschaaf) [reply packages/lit-element/src/lit-element.ts:187]: Background: This is following the patterns already established in `lit-html`

1. Yes, SSR only for LitElement
2. Just extra warning to users that they shouldn't rely on fields in it. Open to other suggestions.
3. We can use a different name; since it's a top-level export, the longer the name the longer

Comment (kevinpschaaf) [reply packages/lit-element/src/lit-element.ts:187]: I updated it to `_$private`

File: rollup-common.js [hunk 8]
```diff
@@ -40,54 +40,75 @@ const skipBundleOutput = {
...
+  _$endNode: 'E',
+  _$getTemplate: 'F',
   // lit-html: TemplateInstance
-  _template: 'H',
+  _$template: 'G',
   // updating-element: UpdatingElement
-  _didUpdate: 'S',
-  _controllers: 'T',
+  _$didUpdate: 'H',
+  _$controllers: 'I',
```
Comment (sorvell) [inline rollup-common.js:78]: Let's try removing this.

**Review by sorvell — APPROVED** [review]
LGTM. (A couple minor nits in my review are being addressed.)

File: packages/lit-element/src/lit-element.ts [hunk 9]
```diff
@@ -170,3 +170,26 @@ if (DEV_MODE) {
...
+ * Private exports for use by other Lit packages, not intended for use by
+ * external users.
+ *
+ * We currently do not make a mangled rollup build of the lit-ssr code. In order
+ * to keep a number of (otherwise private) top-level exports  mangled in the
+ * client side code, we export a _$private object containing those members (or
+ * helper methods for accessing private fields of those members), and then
+ * re-export them for use in lit-ssr. This keeps lit-ssr agnostic to whether the
+ * client-side code is being used in `dev` mode or `prod` mode.
+ */
```
Comment (justinfagnani) [inline packages/lit-element/src/lit-element.ts:186]: might as well add a `@private` jsdoc annotation here for any doc generators

File: packages/lit-element/src/private-ssr-support.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * END USERS SHOULD NOT RELY ON THIS OBJECT.
+ *
+ * We currently do not make a mangled rollup build of the lit-ssr code. In order
+ * to keep a number of (otherwise private) top-level exports  mangled in the
+ * client side code, we export a _$private object containing those members (or
+ * helper methods for accessing private fields of those members), and then
+ * re-export them for use in lit-ssr. This keeps lit-ssr agnostic to whether the
+ * client-side code is being used in `dev` mode or `prod` mode.
```
Comment (justinfagnani) [inline packages/lit-element/src/private-ssr-support.ts:25]: add `@private` jsdoc annotation

File: rollup-common.js [hunk 11]
```diff
@@ -40,54 +40,75 @@ const skipBundleOutput = {
...
+// Private properties which should be stable between versions but are used on
+// unambiguous objects and thus are safe to mangle. These include properties on
+// objects accessed between packages or objects used as values which may be
+// accessed between different versions of a given package.
+//
+// By convention, stable properties should be prefixed with `_$` in the code so
+// they are easily identifiable as properties requiring version stability and
+// thus special attention.
+//
+// Mangled names are uppercase letters, in case we ever might want to use
```
Comment (justinfagnani) [inline rollup-common.js:58]: I think we should consider prefixing with `_` for names of properties that end up on publicly accessible objects, like Parts and TemplateResults.

File: rollup-common.js [hunk 12]
```diff
@@ -40,54 +40,75 @@ const skipBundleOutput = {
...
+  // lit-html: Template (used by platform-support)
+  _$createElement: 'A',
+  _$element: 'B',
+  _$options: 'C',
+  // lit-html: NodePart (used by platform-support)
+  _$startNode: 'D',
+  _$endNode: 'E',
+  _$getTemplate: 'F',
+  // lit-html: TemplateInstance (used by platform-support)
+  _$template: 'G',
```
Comment (justinfagnani) [inline rollup-common.js:75]: I'm surprised TemplateResult.strings and values isn't in here already. We need those to be stable across versions. Can you add them?

**Review by justinfagnani — APPROVED** [review]
LGTM, but we need to add TemplateResult fields.

File: rollup-common.js (hunk 7 above)
Comment (kevinpschaaf) [reply rollup-common.js:65]: Done

File: rollup-common.js (hunk 12 above)
Comment (kevinpschaaf) [reply rollup-common.js:75]: They're not prefixed with `_`, so they don't get mangled, and putting them in this list wouldn't do anything. Do we want them to go ahead and follow the `_$` convention also?

File: rollup-common.js (hunk 11 above)
Comment (kevinpschaaf) [reply rollup-common.js:58]: This is just a list of the "special" mangled properties that need to be stable; there will be other mangled properties on e.g. Parts that will be whatever Terser picks for it. Does this suggest that all Terser mangling should use `_`?

File: packages/lit-element/src/private-ssr-support.ts (hunk 10 above)
Comment (kevinpschaaf) [reply packages/lit-element/src/private-ssr-support.ts:25]: Done

File: rollup-common.js (hunk 12 above)
Comment (justinfagnani) [reply rollup-common.js:75]: Yeah. Removing the `_` was a temporary workaround

File: rollup-common.js (hunk 11 above)
Comment (justinfagnani) [reply rollup-common.js:58]: oh right... we can't control the naming for the other properties. nvm then. If terser ever allows this we can enable it broadly. The trick would be to only incur the extra char on properties of publicly accessible objects.

File: packages/updating-element/src/updating-element.ts [hunk 13]
```diff
@@ -615,7 +615,7 @@ export abstract class UpdatingElement extends HTMLElement {
   /**
    * Set of controllers.
    */
-  _controllers?: Controller[];
+  _$controllers?: Controller[];
```
Comment (sorvell) [inline packages/updating-element/src/updating-element.ts:618]: Take out `_$` from controllers

**Review by sorvell — CHANGES_REQUESTED** [review]

**Review by sorvell — APPROVED** [review]
